Liquity V2 report (#203)
deficollective/defiscan: src/content/protocols/liquity-v2/ethereum.md

---
chain: "Ethereum"
stage: 2
reasons: []
risks: ["L", "L", "L", "L", "L"]
author: ["Rick Pardoe"]
submission_date: "2025-06-07"
publish_date: ""
stage_requirements: [
      [
            { text: "Assets are not in custody by a centralized entity", status: "fixed" },
            { text: "All contracts are verified", status: "fixed" },
            { text: "Source-available codebase", status: "fixed" },
            { text: "Public documentation exists", status: "fixed" },
      ], 
      [
            { text: "Upgrades have no potential of “loss of funds“ ", status: "fixed"},
            { text: "Dependency with a High centralization score is mitigated", status: "fixed" },
            { text: "Frontend backups or self-hosting option exists", status: "fixed"},
      ], 
      [
            { text: "Contracts are immutable", status: "fixed" },
            { text: "Dependency with a High or Medium centralization score is mitigated", status: "fixed" },
            { text: "Alternative third-party frontends exist", status: "fixed" }  
      ]]
---

# Summary

> Liquity v2 is a decentralized borrowing protocol that lets users deposit ETH or LSTs as collateral and mint the stablecoin BOLD. 

Reviewer: We try not to use adjectives that may be subjective, like "decentralized", some more precise alternatives to describe BOLD could be suitable.

```suggestion
Liquity v2 is a borrowing protocol that lets users deposit ETH or Liquid Staking Tokens (LSTs) as collateral and mint the stablecoin `BOLD`. 
```

> BOLD loans need to maintain a minimum collateral ratio of 110% for ETH and 120% for RETH and WSTETH collaterals respectively.
> 
> In addition to the collateral, the loans on each collateral branch are secured by a Stability Pool containing BOLD and by fellow borrowers collectively acting as guarantors of last resort.

Reviewer: ```suggestion
`BOLD` loans need to maintain a minimum collateral ratio of 110% for `ETH` and 120% for `RETH` and `WSTETH` collaterals, respectively.

In addition to the collateral, the loans on each collateral branch are secured by a _Stability Pool_ containing `BOLD` and by fellow borrowers collectively acting as guarantors of last resort.
```

# Ratings

## Chain

The Liquity v2 protocol is deployed on Ethereum mainnet.

> Chain score: Low

## Upgradeability

All permissions in the system have been renounced and the protocol is fully immutable. No upgrades or changes can be made to the protocol, its functions or parameters.

> Upgradeability score: Low

## Autonomy

Liquity v2 relies on a combination of external Chainlink market oracles - ETH-USD, RETH-ETH and STETH-USD - as well as LST smart contract exchange rates in order to price collateral.

If the market oracle or LST exchange rate fails on a given collateral branch, the system shuts that branch down: it freezes new debt issuance, and enables “urgent redemptions” in order to clear debt and collateral as fast as possible. Borrowers are free to repay BOLD, withdraw collateral and close their loans on the shut down branch.

Additionally, if a branch’s LST market oracle fails, the branch falls back to pricing collateral via a combination of the LST exchange rate and ETH-USD price.

> Finally, each collateral branch uses the "last good price" as a last resort if all its price sources become untrusted. This ensures the protocol can continue pricing collateral without interruption, minimizing the impact of oracle failure on its operations. Despite its reliance on external oracles, Liquity v2’s branch shutdown mechanism and fallback pricing logic mitigate the risk of disruption due to price source failure.

Reviewer: Could you please clarify a few points, so that we can ensure this fits a _Low_ centralization score for autonomy:

 - How do you qualify an "Oracle failure", is it simply the Oracle returning 0, or do you also have other measures to prevent Oracle manipulation?
 - As I understand it you use RETH's smart contract exchange rates first, is that also the case for WSTEH or do you use Chainlink's WSTETH-USD primarily?
 - Is the "Last good price" stored in Liquity v2 contracts, or are you referring to Chainlink oracle's last pushed price?
 - Can you define "Urgent redemption" ? Do all positions become redeemable?

Author: Oracle failures 
>How do you qualify an "Oracle failure", is it simply the Oracle returning 0, or do you also have other measures to prevent Oracle manipulation?

V2 deems a Chainlink market oracle to have failed if 1) it returns a price of 0, 2) the call to the oracle reverts or 3) the oracle price has been stale for too long. The staleness thresholds are 24 hours for ETH-USD and STETH-USD (which have update heartbeats of 1 hour) and 48 hours for RETH-ETH (which has an update heartbeat of 24 hours).

Additionally, V2 deems an LST exchange rate to have failed if 1) it returns a value of 0 or 2) the call to the LST contract reverts.

There's no way for any protocol to rigorously detect and prevent all possible oracle manipulation. Whatever thresholds/logic a protocol implements, a malicious actor with technical or economic control over the oracle's price can "fly under the radar" of the protocol's logic/thresholds.

However, V2 still has measures to protect users against oracle manipulation. It uses the "worst" (highest) price for redemptions on WSTETH and RETH branches in order to mitigate adverse redemption arbs, and on the RETH branch, it uses the "worst" (lowest) price for borrowing and liquidations. This mitigates excessive BOLD minting if the RETH market oracle were to be manipulated upwards. This last protection was added since RETH historically has lower liquidity than STETH and market oracle may be more manipulable.

>As I understand it you use RETH's smart contract exchange rates first, is that also the case for WSTEH or do you use Chainlink's WSTETH-USD primarily?

It's a bit more involved - the price calculation depends on both the branch and the operation (i.e. borrowing/liquidation vs redemption). The full logic is laid out here:
https://github.com/liquity/bold?tab=readme-ov-file#choice-of-oracles-and-price-calculations

>Is the "Last good price" stored in Liquity v2 contracts, or are you referring to Chainlink oracle's last pushed price?

Yes, it's stored in v2 contracts. At a given price fetch, the price is calculated according to the logic above - i.e. based on the branch, the user operation and the state of the branch's price sources (i.e. if any have failed or not). That calculated price is then stored as the `lastGoodPrice`.

>Can you define "Urgent redemption" ? Do all positions become redeemable?
Yes, urgent redemptions are a special operation that is enabled once a branch has been shut down (due to oracle failure or insufficient collateralization). With urgent redemptions:
- Any Trove can be directly redeemed from, regardless of interest rate ordering
- No redemption fee is charged
- The redeemer earns a 2% collateral bonus - i.e. for every 1 BOLD redeemed they receive $1.02 worth of collateral

The intent is to clear the debt of the shut down branch as quickly as possible.  Borrowers can also close their Trove at any point after a shut down.

https://github.com/liquity/bold?tab=readme-ov-file#urgent-redemptions

> Autonomy score: Low

## Exit Window

Liquity v2’s contracts are fully immutable - no upgrades or changes can be made, removing the need for an exit window.

## Accessibility

Multiple user interfaces exist and are operated by independent actors ensuring access to the protocol and user funds even if an interface is shutdown or censors a user's transactions.

A list of third-party frontends can be found here: https://www.liquity.org/frontend-v2

> Accessibility score: Low

## Conclusion

> The Liquity v2 deployment on Ethereum Mainnet achieves Low centralization risk score for its Upgradeability, Autonomy, Exit Window and Accessibility dimensions. It thus ranks Stage 2.

Reviewer: ```suggestion
The Liquity v2 deployment on Ethereum mainnet achieves _Low_ centralization risk score for its _Upgradeability_, _Autonomy_, _Exit Window_, and _Accessibility_ dimensions. It thus ranks **Stage 2**.
```

# Protocol Analysis

# Dependencies

The system has the following external dependencies:

> - LST ERC20 tokens: RocketPool ETH (RETH) and Lido Wrapped Staked ETH (WSTETH)
> 
> - Chainlink market oracles: ETH-USD, RETH-ETH and STETH-USD

Reviewer: ```suggestion
- LST ERC20 tokens: RocketPool ETH (`RETH`) and Lido Wrapped Staked ETH (`WSTETH`)

- Chainlink market oracles: ETH-USD, RETH-ETH and STETH-USD
```

Any problem or failure with the collateral contracts could have a significant impact on the Liquity v2 system. However, the system separates different collateral branches and shuts a collateral branch down if the branch collateralization ratio falls too low, or if the LST’s smart contract exchange rate fails. Thus it does its best to mitigate economic or technical collateral failure and remain functional.

Similarly, if a market oracle fails the system shuts down the corresponding branch and falls back to a backup price calculation (see the Autonomy section for details).

In either case of LST exchange rate or market oracle failure, the branch blocks new debt issuance and allows existing users to withdraw their assets.

# Governance

> Liquity v2 incorporates a governance module which gives LQTY stakers voting rights over a fixed portion of the total interest generated from loans. This is termed the “PIL” - Protocol Incentivized Liquidity.
> 
> LQTY stakers vote weekly on initiatives, deciding where to direct that week’s PIL rewards. The intent is to incentivize BOLD liquidity on external markets as per the protocol’s needs.
> 
> LQTY stakers have no control over any other system logic or parameters. Proposing initiatives and claiming an initiative's accrued rewards are completely permissionless.

Reviewer: ```suggestion
Liquity v2 incorporates a governance module which gives LQTY stakers voting rights over a fixed portion of the total interest generated from loans. This is termed the “PIL” - _Protocol Incentivized Liquidity_.

`LQTY` stakers vote weekly on initiatives, deciding where to direct that week’s PIL rewards. The intent is to incentivize `BOLD` liquidity on external markets as per the protocol’s needs.

`LQTY` stakers have no control over any other system logic or parameters. Proposing initiatives and claiming an initiative's accrued rewards are completely permissionless.
```
